feat: add server-side validation to useForm (#4652)
refinedev/refine: documentation/docs/advanced-tutorials/forms/server-side-form-validation.md

Server-side form validation is a technique used to validate form data on the server before processing it. Unlike client-side validation, which is performed in the user's browser using JavaScript, server-side validation occurs on the server-side code, typically in the backend of the application.

## Why Server-Side Validation?

Client-side validation offers a responsive user experience by providing immediate feedback without server round trips. However, it should not be considered a substitute for server-side validation due to its vulnerability to bypassing. Server-side form validation is essential for ensuring data integrity, security, and consistency. It acts as an additional layer that complements client-side validation, preventing malicious or incorrect data from being processed. While client-side validation is valuable, it should not be relied upon exclusively, as server-side validation provides a more robust and reliable validation mechanism.

## Server-Side Validation with **refine**

**refine** supports server-side validation out of the box for all supported UI frameworks (Ant Design, Material UI, Mantine, Chakra UI).

> It requires that the `dataProvider` returning rejected promise with the following shape:

Reviewer: ```suggestion
It expects `dataProvider` to return rejected promise with the following shape:
```

const error: HttpError = {
    message: "Update is not supported in this example.",
    statusCode: 400,
    // the errors field is required for server-side validation.
    // when the errors field is set, useForm will automatically display the error messages in the form with the corresponding fields.
    errors: {
        title: ["Title is required"],
        content: {
            key: "form.error.content",
            message: "Content is required.",
        },
        tags: true,
    },
};
```

> Refer to the `HttpError` type [here][http-error].

`errors` fields can be `string` or `string[]` or `boolean` or `{ key: string; message: string }`

-   `string` or `string[]`: If the field is an array, multiple error messages will be displayed. If the field is a string, only one error message will be displayed.
-   `boolean`: If the field is `true`, "This field is required." message will be displayed. If the field is `false`, no error message will be displayed.
-   `{ key: string; message: string }`: If the field is an object, the `key` field will be used as a translation key. If the `key` is not found in the translation, the `message` field will be displayed.

### How does it work?

> When the `dataProvider` returns rejected promise with [`errors`][http-error] field, the `useForm` hook will set the [`errors`][http-error] state with the `errors` returned from the `dataProvider`.

Reviewer: ```suggestion
When the `dataProvider` returns a rejected promise with [`errors`][http-error] field, `useForm` hook will assign [`errors`][http-error] to the respective form fields.
```
